### ISSUE TITLE
False positive `useless_asref` on `Result`

### ISSUE THREAD

**Issue by domwst** [rust-lang/rust-clippy#16529]
### Summary

When `.as_ref().map(Clone::clone)` is used on `Result<T, E>`, clippy says that it's a useless `as_ref` and suggests replacing `.as_ref().map(...)` with `.clone()` (which changes the expression type from `Result<T, &E>` to `Result<T, E>`). I think rather it should complain about `.map(Clone::clone)` and suggest replacing it with `.cloned()`, which fixes the issue.

### Lint Name

useless_asref

### Reproducer

I tried this code:

```rust
fn main() {
    let r: Result<i32, i64> = Ok(1);
    let _r = r.as_ref().map(Clone::clone);
}
```

I saw this happen:

```
    Checking playground v0.0.1 (/playground)
warning: this call to `as_ref.map(...)` does nothing
 --> src/main.rs:3:14
  |
3 |     let _r = r.as_ref().map(Clone::clone);
  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `r.clone()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.93.0/index.html#useless_asref
  = note: `#[warn(clippy::useless_asref)]` on by default
```

[Link to playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=eec975b6e9aa20a3b1e9f8d7bb88db32)

I expected to see this happen:

No lint or complains about `.map(Clone::clone)`.

### Version

```text
rustc 1.95.0-nightly (f889772d6 2026-02-05)
binary: rustc
commit-hash: f889772d6500faebcac5bb70fa44b5e6581c38cd
commit-date: 2026-02-05
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 22.1.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by mikhailofff** [comment]
I suppose this is a defined behavior. ```map_clone``` does not trigger because of the overlap with ```as_ref```.

https://github.com/rust-lang/rust-clippy/blob/15e0ce9d69340dc83fd8a37ba2c1596354bea6bc/clippy_lints/src/methods/map_clone.rs#L19-L21

**Comment by domwst** [comment]
Yeah, but in this case `useless_asref` doesn't provide better suggestion (in fact it breaks the code).
